### ISSUE TITLE
gatsby-plugin-asset-path causing warm builds to fail

### ISSUE THREAD

**Issue by ezelohar** [gatsbyjs/gatsby#23964]
## Description
When running gatsby build, it fails if cache is not deleted before running build. 

### Steps to reproduce
Clear cache - gatsby clear
Run Build - gatsby build --prefix-paths
Run Build Again - gatsby build --prefix-paths

### Expected result
Not to have errors and to build static version

### Actual result
The code break with 
```
ERROR #98124  WEBPACK

Generating JavaScript bundles failed

Can't resolve '../../../../../public/static/d/3396882757.json' in '/Users/ezelohar/Projects/****/marketing-site/src/components/Molecules/Blog/Social'

Perhaps you need to install the package '../../../../../public/static/d/3396882757.json'?

File: src/components/Molecules/Blog/Social/Social.js
```

### Environment

  System:
    OS: macOS Mojave 10.14.6
    CPU: (8) x64 Intel(R) Core(TM) i7-7820HQ CPU @ 2.90GHz
    Shell: 3.2.57 - /bin/bash
  Binaries:
    Node: 14.1.0 - ~/.nvm/versions/node/v14.1.0/bin/node
    Yarn: 1.10.1 - ~/.yarn/bin/yarn
    npm: 6.14.4 - ~/.nvm/versions/node/v14.1.0/bin/npm
  Languages:
    Python: 2.7.10 - /usr/bin/python
  Browsers:
    Chrome: 83.0.4103.44
    Firefox: 76.0.1
    Safari: 13.1
  npmPackages:
    gatsby: ^2.21.20 => 2.21.20 
    gatsby-image: ^2.4.3 => 2.4.3 
    gatsby-plugin-asset-path: ^2.0.2 => 2.0.2 
    gatsby-plugin-brotli: ^1.3.1 => 1.3.1 
    gatsby-plugin-feed: ^2.5.1 => 2.5.1 
    gatsby-plugin-google-analytics: ^2.3.1 => 2.3.1 
    gatsby-plugin-google-tagmanager: ^2.3.1 => 2.3.1 
    gatsby-plugin-heap: ^1.1.0 => 1.1.0 
    gatsby-plugin-layout: ^1.3.1 => 1.3.1 
    gatsby-plugin-react-helmet: ^3.3.1 => 3.3.1 
    gatsby-plugin-root-import: ^2.0.5 => 2.0.5 
    gatsby-plugin-s3: ^0.3.3 => 0.3.3 
    gatsby-plugin-sharp: ^2.6.2 => 2.6.2 
    gatsby-plugin-sitemap: ^2.4.2 => 2.4.2 
    gatsby-plugin-styled-components: ^3.3.1 => 3.3.1 
    gatsby-plugin-twitter: ^2.3.1 => 2.3.1 
    gatsby-source-contentful: ^2.3.3 => 2.3.3 
    gatsby-source-filesystem: ^2.3.1 => 2.3.1 
    gatsby-transformer-remark: ^2.8.7 => 2.8.7 
    gatsby-transformer-sharp: ^2.5.2 => 2.5.2

**Comment by pvdz** [comment]
@ezelohar from your report it seems like you're running this back-to-back without changing anything else in between, is that correct? No changes locally nor on contentful remotely between these two runs?

**Comment by ezelohar** [comment]
Hi @pvdz 

No changes in between. I just clear cache, run build, it works. Then I just run build again, it fails.
Unfortunately, we use this on gatsby cloud, and because of this, every build will fail until we do manual press on the button and clear cache. 

So I need to find a solution without applying clear cache every time.

**Comment by pvdz** [comment]
Okay that's good.

Looking at your original report, it seems you can repro this locally. I was unable to repro this. I can see your Cloud builds do fail with this but have not been able to manually repro this. But considering it's Cloud, it's not a bug that is bound to a mac or windows env.

I know it's a bit of a tall ask but considering you can repro this locally, is there any chance you could try to get a minimal repro? Would start to try and eliminate contentful from the equation. Then slowly remove plugins until the problem is gone.

I'm happy to double check

I'd also be happy to pair up to minimize this as it's a problem I'd love to resolve. But as I can't repro it locally, it's a bit difficult to reduce to a test case.

**Comment by ezelohar** [comment]
Hi @pvdz 

I removed all plugins except contentful and issue still occur. I can assume it is related to the contentful.

Going further and trying to get it to work without contentful is pointless in my opinion, but if that can help, maybe I can do something.

What do you suggest?

**Comment by pvdz** [comment]
Yeah I understand that. The point would be that I could more easily repro and play with it here, rather than rely on an external data source. But I also understand that it's tedious to get there.

My best suggestion is that we pair up so you can help me poke around in your repro :)

**Comment by mnurik** [comment]
@pvdz @ezelohar I am pretty sure `gatsby-plugin-asset-path` causes that issue, not by contentful plugin. Looking into more details.

**Comment by pvdz** [comment]
Ok, that's a little out of our control. But at least that cuts out the complexity of Contentful from the repro, that's nice. If you can build a repository with locally sourced data that repros this I'm more likely to be able to repro it here as well, and can join in on the fun! :)

**Comment by pvdz** [comment]
I have a repro on another, tiny, site. Also Contentful. Doesn't use that plugin so I expect it to be something else. I'll start digging.

**Comment by pvdz** [comment]
Ah dang, they _are_ using that plugin, just as an inline fork, which was why it was misleading to me.

The plugin is the problem: it's literally moving files from `/public` to somewher else. But `public` and `.cache` are _both_ required for caching to work. So when this plugin is moving `public/static` out, a warm build (when you run `gatsby build` without `gatsby clean` will fail because parts of the cache no longer exists.

The fix for that plugin is to copy rather than move. Or not to use that plugin anymore.

**Comment by pvdz** [comment]
According to the maintainer, @raurir

> Should be fixed in 3.0.1

So please bump your dep and that should fix this issue. Thanks for reporting it!

**Comment by raurir** [comment]
Cheers. Oh, hi mr js1k!
